content/blog/2018-05-23-react-v-16-4.md

```diff
@@ -0,0 +1,120 @@
+---
+title: "React v16.4.0: Experimental Component Profiler and Bugfixes"
+author: [acdlite, bvaughn]
+---
+
+React 16.4 is a minor release that includes an experimental component profiler, as well as a crucial bugfix for `getDerivedStateFromProps`.
+
+## `React.unstable_Profiler`
+
+TODO (@bvaughn)
+
+## Bugfix for `getDerivedStateFromProps`
+
+`getDerivedStateFromProps` is now called every time a component is rendered, regardless of the cause of the update. Previously, it was only called if the component was re-rendered by its parent; it would not fire as the result of a local `setState`. This was an oversight in the initial implementation that has now been corrected. The previous behavior was more similar to how `componentWillReceiveProps` works, but the improved behavior ensures compatibility with React's upcoming asynchronous rendering mode.
+
+**Most likely, this doesn't require any changes to your components**. The rare cases where it does matter fall into one of two categories:
+
+### Avoid Side Effects in `getDerivedStateFromProps`
+
+Like the render method, `getDerivedStateFromProps` should be a pure function of props and state. You should move side-effectful code elsewhere. E.g. Flux dispatches typically belong inside the originating event handler; DOM mutations belong inside `componentDidMount` or `componentDidUpdate`.
+
+Side effects in `getDerivedStateFromProps` were already disallowed, but since it now fires more often than it used to, the recent change may expose previously undiscovered bugs.
+
+### Compare Incoming Props to Previous Props When Computing Controlled Values
+
+The following code assumes `getDerivedStateFromProps` only fires on prop changes. But since it now fires for state changes, too, local updates to the controlled value will be ignored, since the props version always overrides it.
+
+```js
+static getDerivedStateFromProps(props, state) {
+  if (props.value !== state.controlledValue) {
+    return {
+      // Props always win. Oops!
+      controlledValue: props.value,
+    };
+  }
+  return null;
+}
+```
+
+You can change it to do this instead:
+
+```js
+static getDerivedStateFromProps(props, state) {
+  const prevProps = state.prevProps;
+  // Compare incoming props to previous props
+  const controlledValue =
+    prevProps.value !== props.value
+      ? props.value
+      : state.controlledValue;
+  return {
+    // Store the previous props in state
+    prevProps: props,
+    controlledValue,
+  };
+}
+```
```

Comment:
Let's just at least mention that most components **don't** need this pattern. We're only documenting for the cases you're already using it for. But if you just need some value from props, you should just read it from props, and not attempt to put it into state.

Reply:
For example:

>**Note that typically the vast majority of the components don't need `getDerivedStateFromProps` at all.** It shouldn't be used often, as making state dependent on props often leads to logic that is hard to understand. If possible, prefer to have one source of truth for any value (either in props or in state), and [lift state up](https://reactjs.org/docs/lifting-state-up.html) if it's necessary to share it between multiple components.

Reply:
Ok, but I have mixed feelings of telling people not to use a feature, especially one we just added, unless we give them a clear alternatives. We need to follow up with a more comprehensive post soon (like this week or next).